### PR TITLE
Support newer zope.i18n (without actually supporting plural translations)

### LINKED ISSUE #7: Add support for pluralization
zope.i18n 4.5 added support for pluralization, which introduced new methods and new arguments to IMessageCatalog and ITranslationDomain.

zope.app.i18n has local implementations of these interfaces that don't support the new methods/arguments, which breaks zope.app.i18n's test suite.

### PR DESCRIPTION
**Pull request by mgedmin**
Intended to fix #7.

There may be more work to it than I originally thought.

### CODE REVIEW

**Comment by mgedmin** [comment]
Meanwhile I've opened #9 for fixing the tests Right Now to unblock the other PRs.

**Comment by mgedmin** [comment]
I wonder if @thefunny42 who implemented the plural translation support in zope.i18n could be enticed to help with this...

**Comment by thefunny42** [comment]
To be honest we do not use this package, and I am not sure what it does. We use the translation feature with a msgid:
```
         _('This template consists of ${count} item',
            msgid_plural='This template consists of ${count} items',
            number=self.count,
            mapping={'count': self.count})
```

That we use in a template or just call `zope.i18n.translate()` on. The only feedback I have, is that the tool we use to extract translations (a patched lingua to support plurals, among others) does not like at all multiple lines on `msgid_plural`).

**Comment by icemac** [comment]
@mgedmin Is this PR ready to merge? (At least to unblock using a current zope.i18n version.)

**Comment by mgedmin** [comment]
> @mgedmin Is this PR ready to merge?

Not really?

It needs somebody who uses zope.app.i18n to play with the new zope.i18n and see what works and what breaks and what needs fixing.

I'm not currently using zope.app.i18n.

**Comment by icemac** [comment]
Same here: I neither use `zope.app.i18n`.

**Comment by sallner** [comment]
@mgedmin I would actually be very happy, if we could merge that somehow. On multiple occasions I wanted to install `zodbbrowser` lately and there is the transitive dependency via `zope.app.zcmlfiles`, which causes trouble. 

Could it be an option to explicitly raise a `NotImplementedError` if the `translate()` is called with the new attributes and state that in the readme? We could also encourage the users to open up an issue here, in case they need it.

**Comment by mgedmin** [comment]
> Could it be an option to explicitly raise a NotImplementedError if the translate() is called with the new attributes and state that in the readme? We could also encourage the users to open up an issue here, in case they need it.

That sounds better than the status quo, but somebody would have to prepare a PR.

**Comment by sallner** [comment]
Then I will try to do it tomorrow.

**Comment by sallner** [comment]
@mgedmin I think I got every place of use.

File: src/zope/app/i18n/messagecatalog.py [hunk 1]
```diff
@@ -52,10 +52,14 @@ def queryMessage(self, id, default=None):
...
         'See `IMessageCatalog`'
+        # Plural messages are not supported yet. Please report an issue kn
+        # github in case you want to implement this feature.
         raise NotImplementedError
 
     def queryPluralMessage(self, singular, plural, n, dft1=None, dft2=None):
         'See `IMessageCatalog`'
+        # Plural messages are not supported yet. Please report an issue kn
+        # github in case you want to implement this feature.
         raise NotImplementedError
```
Comment (mgedmin) [inline src/zope/app/i18n/messagecatalog.py:63]: Why not make these `raise NotImplemented('plural messages are not supported yet')`?

(Actually, why not include a link to a GitHub ticket in the exception message?)

**Review by mgedmin — COMMENTED** [review]
LGTM

I cannot approve since I opened this PR ;)

File: src/zope/app/i18n/messagecatalog.py (hunk 1 above)
Comment (sallner) [reply src/zope/app/i18n/messagecatalog.py]: I have consolidated the code to only use one error defined in interfaces, including the link.

**Review by mgedmin — COMMENTED** [review]
Nice!

File: CHANGES.rst [hunk 2]
```diff
@@ -11,6 +11,10 @@ CHANGES
   their old locations in zope.component.interfaces instead of their current
   locations in zope.interface.interfaces.
 
+- Add compatibility with ``zope.i18n >= 4.7``. We fulfil the interface for
```
Comment (jamadden) [inline CHANGES.rst:14]:
```suggestion
- Add compatibility with ``zope.i18n >= 4.7``. We fulfill the interface for
```

File: README.rst [hunk 3]
```diff
@@ -1,2 +1,30 @@
+=============
+zope.app.i18n
+=============
+
+.. image:: https://img.shields.io/pypi/v/zope.app.i18n.svg
+        :target: https://pypi.python.org/pypi/zope.app.i18n/
```
Comment (jamadden) [inline README.rst:6]:
```suggestion
        :target: https://pypi.org/project/zope.app.i18n/
```

**Review by jamadden — APPROVED** [review]
Looks fine to me.

**Comment by sallner** [comment]
Thanks for the feedback, I merge that now. @mgedmin Do you mind creating a new release?

**Comment by jamadden** [comment]
4.1.0 is on PyPI.

**Pull request merged**